File: algorithms/cats-and-a-mouse.hs
{-# LANGUAGE FlexibleInstances, UndecidableInstances, DuplicateRecordFields #-}

module Main where

import Control.Monad
import Data.Array
import Data.Bits
import Data.List
import Data.List.Split
import Data.Set
import Debug.Trace
import System.Environment
import System.IO
import System.IO.Unsafe

-- Complete the catAndMouse function below.
catAndMouse :: Int -> Int -> Int -> String
catAndMouse x y z
  | distance (x, z) < distance (y, z) = "Cat A"
  | distance (x, z) > distance (y, z) = "Cat B"
  | otherwise = "Mouse C"
  where distance = abs . uncurry (-)

main :: IO()
main = do
    stdout <- getEnv "OUTPUT_PATH"
    fptr <- openFile stdout WriteMode

    q <- readLn :: IO Int

    forM_ [1..q] $ \q_itr -> do
        xyzTemp <- getLine
        let xyz = words xyzTemp

        let x = read (xyz !! 0) :: Int

        let y = read (xyz !! 1) :: Int

        let z = read (xyz !! 2) :: Int

        let result = catAndMouse x y z

        hPutStrLn fptr result

    hFlush fptr
    hClose fptr

File: hackerrank/algorithms/cats-and-a-mouse.pl
#!/usr/bin/perl

use strict;
use warnings;

# Complete the catAndMouse function below.
sub catAndMouse {
    my ($x, $y, $z) = @_;
    return "Cat A" if abs($x - $z) < abs($y - $z);
    return "Cat B" if abs($x - $z) > abs($y - $z);
    "Mouse C"
}

open(my $fptr, '>', $ENV{'OUTPUT_PATH'});

my $q = <>;
$q =~ s/\s+$//;

for (my $q_itr = 0; $q_itr < $q; $q_itr++) {
    my $xyz = <>;
    $xyz =~ s/\s+$//;
    my @xyz = split /\s+/, $xyz;

    my $x = $xyz[0];
    $x =~ s/\s+$//;

    my $y = $xyz[1];
    $y =~ s/\s+$//;

    my $z = $xyz[2];
    $z =~ s/\s+$//;

    my $result = catAndMouse $x, $y, $z;

    print $fptr "$result\n";
}

close $fptr;
